File: usecase6/model/IH_model_server.py
import umbridge
import numpy as np
import pandas as pd
import os
import argparse
from src.IH_model import IH_powerLaw_stressBased

class IH_Model(umbridge.Model):
    
    _data = None  # Class variable to store shared data
    
    def __init__(self, name):
        self.name = name
        self.params = 3
        if name == "IH_powerLaw_stressBased":
            self.model_func = IH_powerLaw_stressBased

    @classmethod
    def load_data(cls, data_file):
        """Load data once and share across all model instances."""
        if cls._data is None:
            required_columns = ['exposure_time', 'shear_stress']
            df = pd.read_csv(
                data_file,
                usecols = required_columns
                )

            cls._data = df[required_columns].to_numpy()
            print(f"Loaded {data_file} with {len(cls._data)} rows")
        return cls._data

    def get_input_sizes(self, config):
        return [self.params]

    def get_output_sizes(self, config):
        return [len(self._data)]

    def __call__(self, parameters, config):
        try:
            # Parameter validation
            if len(parameters[0]) != self.params:
                raise ValueError(f"Model expects {self.params} parameters, got {len(parameters[0])}")
            
            IH_values = self.evaluate_model(parameters[0])
            return [IH_values.tolist()]
        except Exception as e:
            print(f"Error in model: {e}")
            raise e

    def supports_evaluate(self):
        return True

    def evaluate_model(self, parameters):
        """Evaluate the model with given parameters."""
        A, alpha, beta = parameters

        t_exp_all = self._data[:, 0]
        sigma_all = self._data[:, 1]
        result_all = self.model_func(t_exp_all, sigma_all, A, alpha, beta)
        return result_all

def parse_arguments():
    parser = argparse.ArgumentParser(description='IH Model UMBridge Server')
    parser.add_argument('--name', type=str, 
                       help='Model name')
    parser.add_argument('--data', type=str, 
                       help='Data file path')
    parser.add_argument('--port', type=int, default=4242,
                       help='Server port')
    return parser.parse_args()

def main():
    args = parse_arguments()
    
    # Override with environment variables if present
    name = args.name
    data_file = args.data
    port = args.port
    
    print(f"Starting UMBridge server on port {port}")
    print(f"Data file: {data_file}")
    print(f"Model: {name}")
    
    # Load shared data once
    IH_Model.load_data(data_file)
    
    # Create the model instance
    model = IH_Model(name)
    print(f"Successfully created model: {model.name}")
    
    # Serve the model
    umbridge.serve_models([model], port)

if __name__ == "__main__":
    main()

File: usecase6/model_julia/IH_model_server.jl
using UMBridge
using CSV
using DataFrames
using ArgParse

include("src/IH_model.jl")

function get_IH_model(
    name::String,
    data::DataFrame
    )

    if name=="IH_powerLaw_stressBased"
        model_func = IH_powerLaw_stressBased
        parameters = 3
    elseif name=="IH_powerLaw_strainBased"
        model_func = IH_powerLaw_strainBased_quad
        parameters = 3
    elseif name=="IH_poreFormation_stressBased"
        model_func = IH_poreFormation_stressBased
        parameters = 2
    elseif name=="IH_poreFormation_strainBased"
        model_func = IH_poreFormation_strainBased_quad
        parameters = 2
    end

    # setup umbridge model
    model = UMBridge.Model(
            name = name,
            inputSizes = [parameters],
            outputSizes = [1 for _ in 1:nrow(data)],
            supportsEvaluate = true
        )
    
    function evaluate_model(parameters; data::DataFrame=data)
            t_exp_all = data[:, :exposure_time]
            sigma_all = data[:, :shear_stress]
            return model_func.(t_exp_all, sigma_all, parameters...)
        end


    # evaluation function
    function evaluate(inputs, config)
        IH_values = evaluate_model(inputs[1])
        return [[value] for value in IH_values]
    end

    UMBridge.define_evaluate(model, evaluate)

    return model
end






function parse_arguments()
    s = ArgParseSettings()
    @add_arg_table s begin
        "--name"
            help = "Model name"
            arg_type = String
        "--data"
            help = "Data file path"
            arg_type = String
        "--port"
            help = "Server port"
            arg_type = Int
            default = 49152
    end
    return parse_args(s)
end

function main()
    args = parse_arguments()

    name = args["name"]
    data_file = args["data"]
    port = get(args, "port", 49152)

    println("Starting UMBridge server on port $port")
    println("Data file: $data_file")
    println("Model: $name")

    data = CSV.read(data_file, DataFrame)

    required_columns = ["exposure_time", "shear_stress"]
    if !all(col in names(data) for col in required_columns)
        error("Data file must contain columns: $(join(required_columns, ", "))")
    else 
        select!(data, required_columns)
    end
    
    # Create the model instance
    model = get_IH_model(name, data)

    println("Successfully created model: $(model.name)")

    # Serve the model
    UMBridge.serve_models([model], port)
end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end